File: apps/marketplace/apps/api/src/adapters/push.ts
import { logger } from '../lib/logger.js';

export type PushPayload = {
  to: string | string[];
  title: string;
  body: string;
  data?: Record<string, unknown>;
};

export interface PushAdapter {
  send(payload: PushPayload): Promise<void>;
}

class ExpoPushAdapter implements PushAdapter {
  async send(payload: PushPayload): Promise<void> {
    const tokens = Array.isArray(payload.to) ? payload.to : [payload.to];
    const messages = tokens.map((to) => ({
      to,
      title: payload.title,
      body: payload.body,
      data: payload.data,
      sound: 'default',
    }));
    try {
      const res = await fetch('https://exp.host/--/api/v2/push/send', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(messages),
      });
      if (!res.ok) logger.warn({ status: res.status }, 'Expo push API non-2xx');
    } catch (err) {
      logger.warn({ err }, 'Push send failed (non-fatal)');
    }
  }
}

export const push: PushAdapter = new ExpoPushAdapter();

File: apps/marketplace/apps/api-py/app/adapters/push.py
"""Expo push notification adapter."""
from __future__ import annotations

import logging
from typing import Any

import httpx

log = logging.getLogger(__name__)


async def send_push(
    tokens: list[str],
    title: str,
    body: str,
    data: dict[str, Any] | None = None,
) -> None:
    if not tokens:
        return
    messages = [
        {"to": tok, "title": title, "body": body, "data": data or {}, "sound": "default"}
        for tok in tokens
    ]
    try:
        async with httpx.AsyncClient(timeout=10) as c:
            r = await c.post("https://exp.host/--/api/v2/push/send", json=messages)
        if r.status_code >= 300:
            log.warning("Expo push non-2xx: %s", r.status_code)
    except Exception as exc:  # noqa: BLE001
        log.warning("Push send failed (non-fatal): %s", exc)
